Issue 3 skapa och validera bokningar

joellos opened the pull request:
Add BookingService with CreateBooking flow

Description:
This PR adds the BookingService with core booking operations, including creating, confirming, and retrieving bookings.

Key points:
Booking creation validates required fields, 48h rule, coordinates via MapService, and calculates price via PriceService.

Handles both registered users and guests, with email confirmation for guests.

Private helpers BuildCoordinatesList and BuildPriceCalculationRequest organize service calls.

FluentValidation added for CreateBookingDto.

TaxiSettings issue resolved; price calculation now works. Had to use controller. 

Notes:
Frontend integration not yet included.

Further small, focused commits planned for future improvements.

Make sure the TaxiSettings table is populated before creating a booking; otherwise, price calculation will fail.

The confirmation URL in the emails is currently hardcoded to https://localhost:7161/api/booking/confirm. This means it only works when running the backend locally on that port. Make sure to update it or configure it in appsettings (or via user secrets) if testing on another environment or once the frontend URL is available.

ParreG committed:
Ändrade i booking lite med fler kolumner.

ParreG committed:
Merge branch 'Developer' into issue-74-implement-maps-and-Geocoding

joellos committed:
Merge branch 'Developer' into Issue-3-skapa-och-validera-bokningar

joellos committed:
Merge branch 'Developer' into Issue-3-skapa-och-validera-bokningar

joellos committed:
WIP: Add BookingService with CreateBooking flow and FluentValidation for CreateBookingDto

- Added BookingService with main methods: CreateBookingAsync, ConfirmBookingAsync, GetBookingByIdAsync, GetUserBookingsAsync, etc.
- Implemented private helpers BuildCoordinatesList and BuildPriceCalculationRequest to handle calls to MapService and PriceService.
- CreateBookingAsync flow:
    1. Build list of coordinates and get route info from MapService.
    2. Build price calculation request and get total price from PriceService.
    3. Validate frontend price against calculated price.
    4. Check if booking is for a registered user or guest.
    5. Create booking entity and persist it in the database.
    6. Send appropriate confirmation email (guest or registered user).
- FluentValidation added for CreateBookingDto to ensure data integrity.
- Note: CreateBookingAsync currently fails due to missing TaxiSettings table in DB.

joellos committed:
Never mind what I said earlier about Taxi Settings, It works now. I had to fill the taxi setting using a controller. Fixed confirmationUrl to a local url for now.

joellos committed:
- Removed everything from appsettings and used userSecrets, oops

ParreG reviewed (changes requested):
The booking flow looks good overall, good and strong work, but a few things need to be fixed before merging.

The configuration that is injected is not used, it should be removed or used somewhere.

There should be a check that the trip starts or ends ord have a stop at Arlanda.

IsAvailable should only be true when the booking is confirmed.

If the confirmation token has expired, the booking should be deleted from the database or marked as expired.

I haven’t tested the flow end-to-end yet, I’ll do that tomorrow after these updates are applied.

OssyCA reviewed (commented):
I see that a FluentValidator is created but it's not registered in Program.cs or used manually.

joellos committed:
- Configuration used for ApplicationUrl in email confirmation links
- Arlanda validation checks all stops in trip (pickup, stops, dropoff)
- Flight number required when departing from Arlanda
- Zone pricing validates all stops
- IsAvailable false until booking confirmed
- Expired bookings deleted from database
- Controller updated to expression-bodied pattern matching other controllers
- Auth logic moved to service layer
- GetAvailableBookingsAsync query fixed
- FluentValidation registered in Program.cs

Tested end-to-end:
- Guest booking with email confirmation ✓
- Registered user instant confirmation ✓
- Available bookings query ✓
- Arlanda validation ✓"

ParreG reviewed (changes requested):
The CreateBookingAsync method is well thought out and clearly structured, you handle all key parts of the booking flow: validation, route verification, price calculation, and email sending. The logic makes sense and it covers most edge cases like Arlanda validation and flight number checks. ✅ 

However, the method is too long and tries to handle too many things at once. It mixes validation, data mapping, persistence, and communication logic in a single place, which makes it harder to maintain and test in the future. 

I’d recommend breaking it down into smaller private methods for example: ValidateBookingRequest -> handles time and address rules, ValidateArlandaRules -> checks Arlanda and flight number requirements, CalculateBookingPriceAsync -> calls the price service and validates result, CreateBookingEntity -> builds the Bookings object, HandleBookingEmailAsync -> triggers the email flow. 

That way, the main CreateBookingAsync method only coordinates the process, and each step is easier to test and understand.

Also, the email sending logic should be moved into a separate service (e.g. BookingEmailService or IMailjetService). The BookingService shouldn’t be responsible for building or sending emails directly. For now, I’d suggest removing the mail-building part and just leaving a comment in the code like:
// here we send confirmation mail  
// here we send pending confirmation mail

Overall: great logic, but it needs better separation of concerns to stay scalable as the project grows.

I also noticed that the expectedPrice field is included in the request. This field shouldn’t really be part of backend requests. It’s mainly used by the frontend when the user books a ride, the frontend sends the price shown to the user, and the backend compares it with the calculated price to detect mismatches or manipulation.

joellos committed:
refactor: break down CreateBookingAsync and move email logic to EmailService

- Refactored BookingService.CreateBookingAsync into smaller, focused methods for better scalability and maintainability
- Moved email sending logic from BookingService to EmailService
- Added SendGuestConfirmationEmailAsync and SendRegisteredUserBookingEmailAsync methods to EmailService
- Updated IEmailService interface with new email methods
- Improved separation of concerns between booking logic and email functionality
- Added proper error handling and logging to email methods

Note: Email methods still use MailKit temporarily - ready for Mailjet implementation

joellos committed:
refactor: remove expectedPrice verification from booking creation

- Removed ExpectedPrice property from CreateBookingDto
- Removed price verification logic from BookingService
- Removed ExpectedPrice validation from CreateBookingDTOValidator
- Backend now autonomously calculates and applies the price without client verification

joellos committed:
refactor: major BookingService restructuring and separation of concerns

Breaking Changes:
- Removed ExpectedPrice field from CreateBookingDto
- Backend now autonomously calculates booking prices without client verification

New Services Created:
- BookingValidationService: Handles all validation logic (pickup time, route verification, Arlanda requirements, price calculation)
- BookingFactoryService: Responsible for entity creation, coordinate list building, and price calculation request building
- BookingMapperService: Handles all entity-to-DTO mapping for bookings

Service Improvements:
- BookingService: Reduced from ~500 lines to ~300 lines, now acts as coordinator

Structural Changes:
- Created DTOs/ValidationResults folder for internal validation result classes
  - ValidationResult.cs
  - RouteValidationResult.cs
  - PriceValidationResult.cs
- All booking-related services follow consistent "Service" naming suffix
- Updated DependencyInjections.cs to register all new services

Benefits:
- Improved separation of concerns (Single Responsibility Principle)
- Enhanced testability with smaller, focused services
- Better code maintainability and readability
- Easier to extend with new validation rules or mapping logic
- Clear service boundaries and responsibilities

Updated Files:
- Services/Implementations/BookingService.cs (refactored)
- Services/Implementations/BookingValidationService.cs (new)
- Services/Implementations/BookingFactoryService.cs (new)
- Services/Implementations/BookingMapperService.cs (new)
- Services/Implementations/EmailService.cs (enhanced)
- Services/Interfaces/IBookingValidationService.cs (new)
- Services/Interfaces/IBookingFactoryService.cs (new)
- Services/Interfaces/IBookingMapperService.cs (new)
- Services/Interfaces/IEmailService.cs (updated)
- DTOs/BookingDTOs/CreateBookingDto.cs (removed ExpectedPrice)
- DTOs/ValidationResults/*.cs (new)
- Configurations/DependencyInjections.cs (updated registrations)
- Validators/CreateBookingDTOValidator.cs (removed ExpectedPrice validation)

ParreG reviewed (approved):
Everything works perfectly after testing, great job on this refactor!
The new structure with separate validation, factory, and mapper services makes the whole booking flow much clearer and easier to maintain.

Performance and readability both improved, and the dependency setup feels clean.
No issues found during testing, this is ready to merge.

OssyCA reviewed (approved):
Great job with Booking feature and with all refactoring.
